README.md
# ContainedTabBarView

[![pub package](https://img.shields.io/pub/v/contained_tab_bar_view.svg)](https://pub.dev/packages/contained_tab_bar_view)

`ContainedTabBarView` encapsulates `TabController`, `TabBar` and `TabBarView` into a single, easy to use `Widget` and lets you customize its appearance without worrying about internal workings of "`TabBar` ecosystem".

---

- [Usage](#usage)
    - [Basic Examples](#basic)
    - [Intermediate](#intermediate)
        - [TabBarProperties](#tabbarproperties)
        - [TabBarViewProperties](#tabbarviewproperties)
        - [TabBarPosition](#tabbarposition)
        - [TabBarAlignment](#tabbaralignment)
        - [Intermediate Examples](#intermediate-examples)
    - [Advanced](#advanced)
        - [Change tab from another Widget](#change-tab-from-another-widget)
        - [Advanced Examples](#advanced-examples)
- [Contributing](#contributing)

# Usage

All of the below examples can be found under `example`, each written as a simple `StatelessWidget`.

## Basic

Basic usage includes putting a `ContainedTabBarView` into a `Container` whose `constraints` it will obey. Putting a `ContainedTabBarView` directly into a `body` of a `Scaffold` will, expectedly, draw it full-screen.

```dart
// Example 1
import 'package:contained_tab_bar_view/contained_tab_bar_view.dart';

...

Container(
  padding: const EdgeInsets.all(8.0),
  color: Colors.blue,
  width: 200,
  height: 300,
  child: ContainedTabBarView(
    tabs: [
      Text('First'),
      Text('Second'),
    ],
    views: [
      Container(color: Colors.red),
      Container(color: Colors.green)
    ],
    onChange: (index) => print(index),
  ),
),
```

<p align="center">
<img src="https://raw.githubusercontent.com/slovnicki/contained_tab_bar_view/master/doc/assets/1.gif" alt="example 1" width="220" style="margin-right:16px;margin-bottom:16px">  

<img src="https://raw.githubusercontent.com/slovnicki/contained_tab_bar_view/master/doc/assets/2.gif" alt="example 2" width="220" style="margin-right:16px;margin-bottom:16px">

<img src="https://raw.githubusercontent.com/slovnicki/contained_tab_bar_view/master/doc/assets/3.gif" alt="example 3" width="220" style="margin-right:16px;margin-bottom:16px">
</p>

## Intermediate

### TabBarProperties

All the properties that can be set for a `TabBar` are accessible for setting via `tabBarProperties`. Also, there are many more for the customization of `TabBar`'s appearance. To create your own ("container-based") `indicator`, you can use `ContainerTabIndicator` decoration (see examples 4 and 6) which is a part of this package, but also a separate package whose documentation can be found [here](https://pub.dev/packages/container_tab_indicator). Of course, you can use any other tab bar indicator of your liking.

| Property             | Type                |      Default value        |  Description |
|:---------------------|:--------------------|:--------------------------|:------|
| width                | `double`             | /                         | The width of `TabBar`. If not set, it's full available width. |
| height               | `double `             |  `kToolbarHeight `          | The height of `TabBar`. The `TabBarView` takes the available height minus this value. |
| background           | `Container`           | /                         | Container that is behind the tabs. See example 8. |
| position             | `TabBarPosition`      | `TabBarPosition.top`        | Position of `TabBar` in respect to it's `TabBarView`. See example 5. |
| alignment            | `TabBarAlignment`     | `TabBarAlignment.center`    | Alignment of `TabBar` (if it's width is not full available) within `TabBar`-`TabBarView` flex. See example 5. |
| padding              | `EdgeInsets`          | `const EdgeInsets.all(0.0)` | The `TabBar`s padding. See example 4. |
| margin               | `EdgeInsets`          | `const EdgeInsets.all(0.0)` | The `TabBar`s margin. See example 8. |
| indicator            | `Decoration `         | /                         | The decoration for the active tab. You can use `ContainerTabIndicator` for various customization. See example 4. Also, [see TabBar documentation](https://api.flutter.dev/flutter/material/TabBar/indicator.html) |
| indicatorColor       | `Color `              | /                         | [see TabBar documentation](https://api.flutter.dev/flutter/material/TabBar/indicatorColor.html) |
| indicatorPadding     | `EdgeInsetsGeometry`  | `const EdgeInsets.all(0.0)` | [see TabBar documentation](https://api.flutter.dev/flutter/material/TabBar/indicatorPadding.html) |
| indicatorSize        | `TabBarIndicatorSize` | /                         | [see TabBar documentation](https://api.flutter.dev/flutter/material/TabBar/indicatorSize.html) |
| indicatorWeight      | `double `             | `2.0 `                    | [see TabBar documentation](https://api.flutter.dev/flutter/material/TabBar/indicatorWeight.html) |
| isScrollable         | `bool  `              | `false`                     | [see TabBar documentation](https://api.flutter.dev/flutter/material/TabBar/isScrollable.html) |
| labelColor           | `Color  `             | /                         | [see TabBar documentation](https://api.flutter.dev/flutter/material/TabBar/labelColor.html) |
| labelPadding         | `EdgeInsetsGeometry`  | `const EdgeInsets.all(0.0)` | [see TabBar documentation](https://api.flutter.dev/flutter/material/TabBar/labelPadding.html) |
| labelStyle           | `TextStyle`           | /                         | [see TabBar documentation](https://api.flutter.dev/flutter/material/TabBar/labelStyle.html) |
| unselectedLabelColor | `Color `              | /                         | [see TabBar documentation](https://api.flutter.dev/flutter/material/TabBar/unselectedLabelColor.html) |
| unselectedLabelStyle | `TextStyle`           | /                         | [see TabBar documentation](https://api.flutter.dev/flutter/material/TabBar/unselectedLabelStyle.html) |

### TabBarViewProperties

All the properties that can be set for a `TabBarView` are accessible for setting via `tabBarViewProperties`.

| Property             | Type                |      Default value        |  Description |
|:---------------------|:--------------------|:--------------------------|:------|
| dragStartBehavior                | `DragStartBehavior`             | `DragStartBehavior.start`                         | [See TabBarView documentation](https://api.flutter.dev/flutter/material/TabBarView/dragStartBehavior.html) |
| physics               | `ScrollPhysics `             |  (platform specific)          | [See TabBarView documentation](https://api.flutter.dev/flutter/material/TabBarView/physics.html) |


For example, if you want to disable swiping behavior on the `TabBarView`, set the option like this:
```dart
ContainedTabBarView(
  ...
  tabBarViewProperties: TabBarViewProperties(
    physics: NeverScrollableScrollPhysics(),
  ),
)
```


### TabBarPosition

```dart
enum TabBarPosition { top, bottom, left, right }
```

### TabBarAlignment

```dart
enum TabBarAlignment { start, center, end }
```

### Intermediate Examples

```dart
// Example 5
ContainedTabBarView(
  tabs: [
      Text('First'),
      Text('Second'),
  ],
  tabBarProperties: TabBarProperties(
    width: 200,
    height: 32,
    background: Container(
      decoration: BoxDecoration(
        color: Colors.blue,
        borderRadius: BorderRadius.all(Radius.circular(8.0)),
        boxShadow: [
          BoxShadow(
            color: Colors.black.withOpacity(0.4),
            spreadRadius: 0.5,
            blurRadius: 2,
            offset: Offset(1, -1),
          ),
        ],
      ),
    ),
    position: TabBarPosition.bottom,
    alignment: TabBarAlignment.end,
    indicatorColor: Colors.transparent,
    labelColor: Colors.white,
    unselectedLabelColor: Colors.grey[400],
  ),
  views: [
    Container(color: Colors.red),
    Container(color: Colors.green),
  ],
)
```

<p align="center">
<img src="https://raw.githubusercontent.com/slovnicki/contained_tab_bar_view/master/doc/assets/4.gif" alt="example 4" width="220" style="margin-right:16px;margin-bottom:16px">

<img src="https://raw.githubusercontent.com/slovnicki/contained_tab_bar_view/master/doc/assets/5.gif" alt="example 5" width="220" style="margin-right:16px;margin-bottom:16px">

<img src="https://raw.githubusercontent.com/slovnicki/contained_tab_bar_view/master/doc/assets/6.gif" alt="example 6" width="220" style="margin-right:16px;margin-bottom:16px">
</p>

## Advanced

### Change tab from another Widget

To control the state, i.e. move to another tab, from outside of `ContainedTabBarView` create it with a `GlobalKey` and then use methods exposed from state: `animateTo()`, `next()` and  `previous()`.

```dart
// Example 7
class Example7 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    GlobalKey<ContainedTabBarViewState> _key = GlobalKey();
    ContainedTabBarView containedTabBarView = ContainedTabBarView(
      key: _key,
      tabs: [
        Text('First'),
        Text('Second'),
      ],
      views: [
        Container(color: Colors.red),
        Container(color: Colors.green),
      ],
    );

    return Scaffold(
      appBar: AppBar(
        title: Text('Example 7'),
      ),
      body: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          ElevatedButton(
            onPressed: () => _key.currentState?.previous(),
            child: Icon(Icons.arrow_back_ios),
          ),
          Center(
            child: Container(
              padding: const EdgeInsets.all(8.0),
              color: Colors.blue,
              width: 200,
              height: 300,
              child: containedTabBarView,
            ),
          ),
          ElevatedButton(    
            onPressed: () => _key.currentState?.next(),
            child: Icon(Icons.arrow_forward_ios),
          ),
        ],
      ),
    );
  }
}
```

### Advanced Examples

<p align="center">
<img src="https://raw.githubusercontent.com/slovnicki/contained_tab_bar_view/master/doc/assets/7.gif" alt="example 7" width="220" style="margin-right:16px;margin-bottom:16px">

<img src="https://raw.githubusercontent.com/slovnicki/contained_tab_bar_view/master/doc/assets/8.gif" alt="example 8" width="220" style="margin-right:16px;margin-bottom:16px">

<img src="https://raw.githubusercontent.com/slovnicki/contained_tab_bar_view/master/doc/assets/9.gif" alt="example 9" width="220" style="margin-right:16px;margin-bottom:16px">
</p>

# Contributing

This package is still in early stages. If you notice any bugs, please file an [issue](https://github.com/slovnicki/contained_tab_bar_view/issues). If you are willing to fix or enhance things yourself, you are very welcome to make a pull request.

CHANGELOG.md
# 0.8.0

- **Breaking** (`TabBarProperties`): Rename `innerPadding` to `padding` and `outerPadding` to `margin`
- **Migrate** to null safety
- **Add** doc comments

# 0.7.1

- fix calling `onChange` twice when tab is tapped

# 0.7.0

- add `tabBarViewProperties`
- add `innerPadding` option to `tabBarProperties`

# 0.6.0

- export `container_tab_indicator` package as a part of this package
- leave `ContainerIndicator` for now, but will be removed in the future

# 0.5.0

- add `outerPadding` option to `TabBarProperties`
- change `backgroundStack` to `background` (`Container`)

# 0.4.0

- expose some `State` methods for switching tabs from outside
- enable the usage of `BorderRadius` at `ContainerIndicator`, instead of just `double`
- move `tabBarPosition` and `tabBarAlignment` into `TabBarProperties`
- rename `backgroundDecoration` to `backgroundStack` and enhance usage possibilities to arbitrary `Stack` instead of just `BoxDecoration`
- implement `TabBarPosition.left` and `TabBarPosition.right`

# 0.3.2

- format code with `dartfmt`

# 0.3.1

- add `ContainerIndicator`

# 0.3.0

- add `tabBarAlignment` and `TabBarProperties.backgroundDecoration` options

# 0.2.0

- implement `tabBarProperties` option

# 0.1.0

- `ContainedTabBarView` can be placed as a chid of `Container`, without causing overflows.
- `onChange` works


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:contained_tab_bar_view/contained_tab_bar_view.dart';

void main() => runApp(App());

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text('ContainedTabBarView Example'),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              ExampleButton(route: '/example1'),
              ExampleButton(route: '/example2'),
              ExampleButton(route: '/example3'),
              ExampleButton(route: '/example4'),
              ExampleButton(route: '/example5'),
              ExampleButton(route: '/example6'),
              ExampleButton(route: '/example7'),
              ExampleButton(route: '/example8'),
              ExampleButton(route: '/example9'),
              ExampleButton(route: '/example10'),
            ],
          ),
        ),
      ),
      routes: {
        '/example1': (context) => Example1(),
        '/example2': (context) => Example2(),
        '/example3': (context) => Example3(),
        '/example4': (context) => Example4(),
        '/example5': (context) => Example5(),
        '/example6': (context) => Example6(),
        '/example7': (context) => Example7(),
        '/example8': (context) => Example8(),
        '/example9': (context) => Example9(),
        '/example10': (context) => Example10(),
      },
    );
  }
}

class ExampleButton extends StatelessWidget {
  ExampleButton({required this.route});
  final String route;

  @override
  Widget build(BuildContext context) {
    return ElevatedButton(
      onPressed: () => Navigator.of(context).pushNamed(route),
      child: Text(
        route,
        style: TextStyle(color: Colors.white),
      ),
    );
  }
}

class Example1 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Example 1'),
      ),
      body: Center(
        child: Container(
          padding: const EdgeInsets.all(8.0),
          color: Colors.blue,
          width: 200,
          height: 300,
          child: ContainedTabBarView(
            tabs: [
              Text('First'),
              Text('Second'),
            ],
            views: [
              Container(color: Colors.red),
              Container(color: Colors.green)
            ],
            onChange: (index) => print(index),
          ),
        ),
      ),
    );
  }
}

class Example2 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Example 2'),
      ),
      body: ContainedTabBarView(
        tabs: [
          Text('First', style: TextStyle(color: Colors.black)),
          Text('Second', style: TextStyle(color: Colors.black))
        ],
        views: [
          Container(color: Colors.red),
          Container(color: Colors.green),
        ],
        onChange: (index) => print(index),
      ),
    );
  }
}

class Example3 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Example 3'),
      ),
      body: ContainedTabBarView(
        tabs: [
          Text('First'),
          Text('Second'),
        ],
        tabBarProperties: TabBarProperties(
          height: 32.0,
          indicatorColor: Colors.black,
          indicatorWeight: 6.0,
          labelColor: Colors.black,
          unselectedLabelColor: Colors.grey[400],
        ),
        views: [
          Container(color: Colors.red),
          Container(color: Colors.green),
        ],
        onChange: (index) => print(index),
      ),
    );
  }
}

class Example4 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Example 4'),
      ),
      body: ContainedTabBarView(
        tabs: [
          Text('First'),
          Text('Second'),
        ],
        tabBarProperties: TabBarProperties(
          padding: const EdgeInsets.symmetric(
            horizontal: 32.0,
            vertical: 8.0,
          ),
          indicator: ContainerTabIndicator(
            radius: BorderRadius.circular(16.0),
            color: Colors.blue,
            borderWidth: 2.0,
            borderColor: Colors.black,
          ),
          labelColor: Colors.white,
          unselectedLabelColor: Colors.grey[400],
        ),
        views: [
          Container(color: Colors.red),
          Container(color: Colors.green),
        ],
        onChange: (index) => print(index),
      ),
    );
  }
}

class Example5 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Example 5'),
      ),
      body: Center(
        child: Container(
          height: 300,
          width: 360,
          child: ContainedTabBarView(
            tabs: [
              Text('First'),
              Text('Second'),
            ],
            tabBarProperties: TabBarProperties(
              width: 200,
              height: 32,
              background: Container(
                decoration: BoxDecoration(
                  color: Colors.blue,
                  borderRadius: BorderRadius.all(Radius.circular(8.0)),
                  boxShadow: [
                    BoxShadow(
                      color: Colors.black.withOpacity(0.4),
                      spreadRadius: 0.5,
                      blurRadius: 2,
                      offset: Offset(1, -1),
                    ),
                  ],
                ),
              ),
              position: TabBarPosition.bottom,
              alignment: TabBarAlignment.end,
              indicatorColor: Colors.transparent,
              labelColor: Colors.white,
              unselectedLabelColor: Colors.grey[400],
            ),
            views: [
              Container(color: Colors.red),
              Container(color: Colors.green)
            ],
            onChange: (index) => print(index),
          ),
        ),
      ),
    );
  }
}

class Example6 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Example 6'),
      ),
      body: Center(
        child: Container(
          width: 300,
          height: 200,
          child: ContainedTabBarView(
            tabs: [
              ...[1, 2]
                  .map(
                    (e) => Container(
                      width: 16,
                      height: 16,
                      decoration: BoxDecoration(
                        color: Colors.transparent,
                        border: Border.all(color: Colors.grey[600]!),
                        borderRadius: BorderRadius.all(
                          Radius.circular(4.0),
                        ),
                      ),
                    ),
                  )
                  .toList(),
            ],
            tabBarProperties: TabBarProperties(
              width: 64,
              height: 32,
              position: TabBarPosition.bottom,
              indicator: ContainerTabIndicator(
                width: 16.0,
                height: 16.0,
                radius: BorderRadius.circular(4.0),
                color: Colors.blue,
              ),
            ),
            views: [
              Container(color: Colors.red),
              Container(color: Colors.green)
            ],
            onChange: (index) => print(index),
          ),
        ),
      ),
    );
  }
}

class Example7 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    GlobalKey<ContainedTabBarViewState> _key = GlobalKey();
    ContainedTabBarView containedTabBarView = ContainedTabBarView(
      key: _key,
      tabs: [
        Text('First'),
        Text('Second'),
      ],
      views: [
        Container(color: Colors.red),
        Container(color: Colors.green),
      ],
      onChange: (index) => print(index),
    );

    return Scaffold(
      appBar: AppBar(
        title: Text('Example 7'),
      ),
      body: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          ElevatedButton(
            onPressed: () => _key.currentState?.previous(),
            child: Icon(Icons.arrow_back_ios),
          ),
          Center(
            child: Container(
              padding: const EdgeInsets.all(8.0),
              color: Colors.blue,
              width: 200,
              height: 300,
              child: containedTabBarView,
            ),
          ),
          ElevatedButton(
            child: Icon(Icons.arrow_forward_ios),
            onPressed: () => _key.currentState?.next(),
          ),
        ],
      ),
    );
  }
}

class Example8 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Example 8'),
      ),
      body: Center(
        child: Container(
          width: 200,
          height: 300,
          child: ContainedTabBarView(
            tabs: [
              Text('First'),
              Text('Second'),
            ],
            tabBarProperties: TabBarProperties(
              margin: EdgeInsets.only(bottom: 8.0),
              background: Container(
                decoration: BoxDecoration(
                  color: Colors.blue,
                  borderRadius: BorderRadius.all(Radius.circular(8.0)),
                  boxShadow: [
                    BoxShadow(
                      color: Colors.black.withOpacity(0.4),
                      spreadRadius: 0.5,
                      blurRadius: 2,
                      offset: Offset(1, -1),
                    ),
                  ],
                ),
                child: Center(
                  child: Container(
                    width: 16,
                    height: 16,
                    decoration: BoxDecoration(
                      color: Colors.blue[700],
                      borderRadius: BorderRadius.all(Radius.circular(8.0)),
                    ),
                  ),
                ),
              ),
            ),
            views: [
              Container(color: Colors.red),
              Container(color: Colors.green)
            ],
            onChange: (index) => print(index),
          ),
        ),
      ),
    );
  }
}

class Example9 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Example 9'),
      ),
      body: Center(
        child: Container(
          width: 300,
          height: 200,
          child: ContainedTabBarView(
            tabs: [
              ...[1, 2]
                  .map(
                    (e) => Container(
                      width: 16,
                      height: 16,
                      decoration: BoxDecoration(
                        color: Colors.transparent,
                        border: Border.all(color: Colors.grey[600]!),
                        borderRadius: BorderRadius.all(
                          Radius.circular(8.0),
                        ),
                      ),
                    ),
                  )
                  .toList()
            ],
            tabBarProperties: TabBarProperties(
              height: 48,
              width: 48,
              position: TabBarPosition.right,
              indicator: ContainerTabIndicator(
                width: 16.0,
                height: 16.0,
                radius: BorderRadius.circular(8.0),
                color: Colors.blue,
              ),
            ),
            views: [
              Container(color: Colors.red),
              Container(color: Colors.green)
            ],
            onChange: (index) => print(index),
          ),
        ),
      ),
    );
  }
}

class Example10 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Example 10'),
      ),
      body: Center(
        child: Container(
          padding: const EdgeInsets.all(8.0),
          color: Colors.blue,
          width: 200,
          height: 300,
          child: ContainedTabBarView(
            tabs: [
              Text('First'),
              Text('Second'),
            ],
            views: [
              Container(color: Colors.red),
              Container(color: Colors.green)
            ],
            tabBarViewProperties: TabBarViewProperties(
              physics: NeverScrollableScrollPhysics(),
            ),
            onChange: (index) => print(index),
          ),
        ),
      ),
    );
  }
}
